Fix formatting string for assembly MVID mismatch

trylek opened the pull request:
During development of the MVID check change I at some point ended
up with mismatched printf specifiers and arguments - the string
is Unicode so that we need %S to print the 8-bit string
returned from GetSimpleName.

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix formatting string for assembly MVID mismatch

During development of the MVID check change I at some point ended
up with mismatched printf specifiers and arguments - the string
is Unicode so that we need %S to print the 8-bit string
returned from GetSimpleName.

Thanks

Tomas

janvorli commented on src/coreclr/vm/nativeimage.cpp:
```diff
@@ -285,7 +285,7 @@ void NativeImage::CheckAssemblyMvid(Assembly *assembly) const
 
     SString message;
-    message.Printf(W("MVID mismatch between loaded assembly '%s' (MVID = %s) and an assembly with the same simple name embedded in the native image '%s' (MVID = %s)"),
+    message.Printf(W("MVID mismatch between loaded assembly '%S' (MVID = %s) and an assembly with the same simple name embedded in the native image '%s' (MVID = %s)"),
```
It seems we could also change the native image formatting to `%S` and replace the `SString(SString::Utf8, GetFileName()).GetUnicode()` by just `GetFileName()`. Or maybe the other way - keep the `%s` where you've changed it to `%S` and change the related argument to `SString(SString::Utf8, assembly->GetSimpleName()).GetUnicode()`.  
The reason I am suggesting the other way is that I am not sure if message.Printf can handle UTF-8 strings (maybe it would consider the 8 bit strings to be encoded using something like Windows 1250). Can you please test it with some assembly name that would have non-ascii characters in the name to make sure it handles them properly?
The simpler of way without re-encoding to UTF16 would be preferred if it works.

trylek replied:
Thanks Jan for pointing that out, I have verified locally that Printf indeed prints complete nonsense when presented with a UTF-8 string. I have modified the code to use Unicode per your suggestion. Based on our offline discussion I also had to update the <code>CrstTypes.def</code> file to declare the fact that the <code>AssemblyLoader</code> lock may now also need to take <code>DebuggerMutex</code> in the case when the MVID validation fails and the application fails fast.

trylek committed:
Jan's PR feedback; AssemblyLoader vs. DebuggerMutex lock ordering

jkotas commented on src/coreclr/vm/nativeimage.cpp:
```diff
@@ -286,7 +286,7 @@ void NativeImage::CheckAssemblyMvid(Assembly *assembly) const
     SString message;
     message.Printf(W("MVID mismatch between loaded assembly '%s' (MVID = %s) and an assembly with the same simple name embedded in the native image '%s' (MVID = %s)"),
-        assembly->GetSimpleName(),
+        SString(SString::Utf8, assembly->GetSimpleName()).GetUnicode(),
```
When is the destructor of the `SString` object that holds the string going to be called? 

It has be called after `Printf` finishes. Otherwise, the pointer returned from `GetUnicode` is going to point to freed memory. I do not know whether C++ standard guarantees it or not.

trylek replied:
I believe that the C++ standard does guarantee that all such destructions happen at the end of the "statement", in this particular case the expression statement calling the <code>Printf</code> function. I'll try to look up the relevant reference in the C++ standard. Alternatively, I can easily make these two local variables if you think it's safer or if it's more in line with typical CoreCLR runtime coding style.

trylek replied:
While the official C++ standard is not freely available, at least the cppreference web page states in the section

https://en.cppreference.com/w/cpp/language/implicit_conversion#Temporary_materialization

that:

*All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation. This is true even if that evaluation ends in throwing an exception.

There are two exceptions from that:

The lifetime of a temporary object may be extended by binding to a const lvalue reference or to an rvalue reference (since C++11), see reference initialization for details.
The lifetime of a temporary object created when evaluating the default arguments of a default or copy constructor used to initialize an element of an array ends before the next element of the array begins initialization.*

jkotas replied:
Sounds good. Thanks for checking.

trylek commented:
The field offset verification fix is still hitting issues on ARM and will likely take a few more hours to land; for this reason I'm retesting this PR in the hope that after JanK's integration of the NullRef fix it starts passing and can be merged in immediately.

trylek commented:
I believe I have addressed all PR feedback and the latest run hasn't (yet) uncovered any new issues, the single-file leg seems affected by a network failure. Can you please take another look when you have a chance and let me know if you think the change is good to go in and start untangling our various CG2 pending issues?

jkotas reviewed (approved):
LGTM. Thanks!
